provisioner/openid_provider.go

```diff
@@ -0,0 +1,158 @@
+package provisioner
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	stsDomain = "sts.amazonaws.com"
+	// acmRootCAThumbprint is the SHA-1 sum of the root certificate in the trust chain for the certificate use to serve
+	// the open id discovery document.
+	acmRootCAThumbprint = "9e99a48a9960b14926bb7f3b02e22da2b0ab7280"
+)
+
+type OpenIDProviderProvisioner struct {
+	awsAdapter  *awsAdapter
+	providerUrl string
+	clientIDs   []string
+	thumbprints []string
+}
+
+func NewOpenIDProviderProvisioner(adapter *awsAdapter, providerHostname string) (*OpenIDProviderProvisioner, error) {
+	return &OpenIDProviderProvisioner{
+		awsAdapter:  adapter,
+		providerUrl: providerHostname,
+		clientIDs:   []string{stsDomain},
+		thumbprints: []string{acmRootCAThumbprint},
+	}, nil
+}
+
+func (p *OpenIDProviderProvisioner) Provision() error {
+	providers, err := p.awsAdapter.iamClient.ListOpenIDConnectProviders(&iam.ListOpenIDConnectProvidersInput{})
+	if err != nil {
+		return fmt.Errorf("failed to list openid providers: %v", err)
+	}
+	var existingProvider *iam.GetOpenIDConnectProviderOutput
+	var existingProviderArn string
+	for _, prv := range providers.OpenIDConnectProviderList {
+		provider, err := p.awsAdapter.iamClient.GetOpenIDConnectProvider(&iam.GetOpenIDConnectProviderInput{OpenIDConnectProviderArn: prv.Arn})
+		if err != nil {
+			return fmt.Errorf("failed to get openid provider: %v", err)
+		}
+		if *provider.Url == p.providerUrl {
+			existingProvider = provider
+			existingProviderArn = *prv.Arn
+		}
+	}
+
+	if existingProvider != nil {
+		if !stringArrayContains(aws.StringValueSlice(existingProvider.ThumbprintList), p.thumbprints) {
```

Comment:
Same question here: wouldn't it be easier to just overwrite the list with the expected value?

Reply:
Again rotation of certificates